vllm-ext/vllm/extension/ns/model/ns_model.py

```diff
@@ -0,0 +1,324 @@
+from vllm.model_executor.layers.linear import LinearMethodBase
+from vllm.attention import AttentionMetadata
+
+from inference_engine import Model as IE_Model
+
+from vllm.model_executor.layers.quantization.base_config import QuantizationConfig
+from vllm.model_executor.layers.vocab_parallel_embedding import VocabParallelEmbedding
+from vllm.sequence import SamplerOutput, SequenceGroupMetadata
+
+from vllm.logger import init_logger
+
+from vllm.extension import ns
+
+from vllm.extension.ns.model import _ModelPerf
+
+
+# import threading
+
+logger = init_logger(__name__)
+
+_NS_NUM_THREADS = "NS_NUM_THREADS"  # threads for prompt decoding
+_NS_NUM_THREADS_NEXT = "NS_NUM_THREADS_NEXT"  # threads for next token generation
+_NS_MAX_PROMPT_TOKENS = "NS_MAX_PROMPT_TOKENS"  # to be supported
+
+
+class NSModel(nn.Module):
+    def __init__(
+        self,
+        config: PretrainedConfig,
+        cache_config: Optional[CacheConfig] = None,
+        quant_config: Optional[QuantizationConfig] = None,
+        lora_config: Optional[LoRAConfig] = None,
+    ):
+        super(NSModel, self).__init__()
+        self.config = config
+        self.quant_config = quant_config
+        self.lora_config = lora_config
+        self.ie_model = None
+        self.tokenizer = None
+
+        self.padding_idx = config.pad_token_id
+        lora_vocab = (
+            (lora_config.lora_extra_vocab_size * (lora_config.max_loras or 1)) if lora_config else 0
+        )
+        self.vocab_size = config.vocab_size + lora_vocab
+        self.org_vocab_size = config.vocab_size
+        self.embed_tokens = VocabParallelEmbedding(
+            self.vocab_size,
+            config.hidden_size,
+            org_num_embeddings=config.vocab_size,
+        )
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        positions: torch.Tensor,
+        kv_caches: List[torch.Tensor],
+        attn_metadata: AttentionMetadata,
+    ):
+        assert len(kv_caches) == 1, "kv_caches should have 1 element here"
+        # use data_ptr and torch type in str to avoid inference engine model depends on pytorch and vllm types
+        # kv_cache pointer is set in NSBlockManager in advance
+        return self.ie_model(
+            input_ids.data_ptr(),
+            positions.data_ptr(),
+            attn_metadata.is_prompt,
+            attn_metadata.block_tables.data_ptr(),
+            attn_metadata.slot_mapping.data_ptr(),
+            attn_metadata.seq_lens,
+        )
+
+    def init_inference_engine(
+        self,
+        model_config: ModelConfig,
+        parallel_config: ParallelConfig,
+        scheduler_config: SchedulerConfig,
+    ):
+        # get available cores
+        try:
+            max_prompt_tokens = int(os.environ.get(_NS_MAX_PROMPT_TOKENS, "8192"))
+            # cpus_per_work is set to 1 for better ns perf so it's inappropriate to use ray to get available cores
```

Comment:
you can directly use `cpus_per_work` to set cpu cores for ns, no need to set cpus_per_work to 1 and use another env for ns.

Reply:
no, we cannot. cpus_per_worker value is set to OMP_NUM_THREADS. I actually need to set OMP_NUM_THREADS to 1 to get better perf.